src/beacons/BeaconsContext.tsx

```diff
@@ -97,36 +86,54 @@ enum storeKey {
 const BeaconsContext = createContext<BeaconsContextState>(defaultState);
 
 const BeaconsContextProvider: React.FC = ({children}) => {
+  const {rationaleMessages} = useBeaconsMessages();
+  const [beaconsInfo, setBeaconsInfo] = useState<BeaconsInfo>();
+  const [isConsentGranted, setIsConsentGranted] = useState<boolean>(false);
   const [isBeaconsEnabled, debugOverrideReady] = useIsBeaconsEnabled();
+
+  const isInitializedRef = useRef(false);
   const isBeaconsSupported =
     isBeaconsEnabled && debugOverrideReady && !!KETTLE_API_KEY;
-  const {rationaleMessages} = useBeaconsMessages();
-  const isInitializedRef = useRef(false);
-  const [beaconsInfo, setBeaconsInfo] = useState<BeaconsInfo>();
 
   const updateBeaconsInfo = () => getBeaconsInfo().then(setBeaconsInfo);
 
-  const initializeKettleSDK = useCallback(async () => {
-    if (!isInitializedRef.current) {
+  const initializeKettleSDK = useCallback(
+    async (bypassPermissions: boolean) => {
+      // By checking permissions we can avoid initializing the SDK if the user
+      // have not granted any permissions. This is to avoid the SDK to start
+      // collecting data without the user knowing.
       const permissions = await allowedPermissionsForBeacons();
-      if (permissions.length > 0) {
+      if (
+        !isInitializedRef.current &&
+        (permissions.length > 0 || bypassPermissions)
+      ) {
         await NativeModules.KettleSDKExtension.initializeKettleSDK();
         isInitializedRef.current = true;
       }
+    },
+    [],
+  );
+
+  const stopBeacons = useCallback(async () => {
+    const permissions = await allowedPermissionsForBeacons();
+    if (
+      permissions.length > 0 &&
+      beaconsInfo?.isStarted &&
+      isInitializedRef.current
+    ) {
+      Kettle.stop(permissions);
```

Comment:
How important is it that this list of permissions when stopping matches the list it was started with?

From the looks of it, it seems that we fetch permissions from the OS layer before stopping, and it might happen that someone has changed OS permissions between starting and stopping the Kettle module.

Reply:
I have changed this to reflect the changes I made in the stop, this stop now should stop all as it was fixed for the auto-stopping.